internal/broker/broker.go

```diff
@@ -636,6 +644,25 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 	return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 }
 
+// userNameIsAllowed checks whether the user's username is allowed to access the machine.
+func (b *Broker) userNameIsAllowed(userName string) bool {
+	// The user is allowed to log in if:
+	// - ALL users are allowed
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine and OWNER is in the allowed_users list
+	if b.cfg.userConfig.allUsersAllowed {
+		return true
+	}
+	if _, ok := b.cfg.userConfig.allowedUsers[userName]; ok {
+		return true
+	}
+	if !b.cfg.userConfig.ownerAllowed {
+		return false
+	}
+	// If owner is undefined, then the first user to login is considered the owner
+	return b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName
```

Comment:
There's a race condition when two users log in at the same time. I tried to fix that in 9cc259e0d8bc26e125009b4d934292cbdcbdbc55, feel free to cherry-pick.

Reply:
@adombeck are you able to provide a test for it?

at the same time being conscious of the previous use case where all users are allowed, this is a new functionality that can simply go in a separate PR and we can write a `TODO` that is going to get addressed

Reply:
In my opinion we don't necessarily need a test for it, and because it seems to require quite some effort to add one, I'd say lets merge it without the test.